web/bb-webpage.cgi.1
.TH BB-WEBPAGE 1 "Version 4.2.1: 27 Nov 2008" "Xymon"
.SH NAME
bb-webpage \- Utility to show a webpage using header and footer
.SH SYNOPSIS
.B "bb-webpage [options]"

.SH DESCRIPTION
\fBbb-webpage\fR is a tool to generate a webpage in the Hobbit style,
with a standard header- and footer as well as a Hobbit background.
The data to present on the webpage, apart from the header and
footer, are passed to bb-webpage in stdin. The generated webpage
is printed to stdout.

.SH OPTIONS
.IP "--env=FILENAME"
Loads the environment defined in FILENAME before executing the CGI script.

.IP "--hffile=PREFIX"
Use the header- and footer-files in $BBHOME/web/PREFIX_header and PREFIX_footer.
If not specified, bb_header and bb_footer are used.

.IP "--color=COLOR"
Set the background color of the generated webpage to COLOR. Default: Blue

.IP "--debug"
Enable debugging output.

.SH "SEE ALSO"
hobbit(7)



Changes
Changes from 4.3.0 beta 1 -> 4.3.0 beta 2 (24 Apr 2009)
=======================================================
* New "--shuffle" option for bbtest-net to run network tests
  in a random order.
* Client startup script now exports important environment variables,
  so they are actually used in systems with a traditional shell.
* hobbitlaunch no longer crashes if there are no tasks
* Client configure script includes librt for clock_gettime()
* New client support for mainframes: z/OS, z/VM, z/VSE
* Enhanced eventlog and top-changing hosts webpages.
* Revert debian package pathnames back to use "hobbit", so
  updates from 4.2.x will actually work.
* Ghostlist options in hobbitcgi.cfg had no effect because
  of typo in setting name.
* "data" messages could crash hobbitd.
* Debug output from hobbitd_channel now logs only the relevant
  data instead of the full message.
* trimhistory now informs the history module to re-open the
  "allevents" file after trimming it.
* devmon template fix
* New "rrdcachectl" utility included.
* Fixed sorting routine (affected holiday list and others)
* Fix generic crash in communications module between Xymon
  programs, where an empty response message would crash caller.

Changes from 4.2.3 -> 4.3.0 beta 1 (09 Feb 2009)
================================================
Core changes:
* New API's for loadhosts and sendmessage, in preparation for
  the full 5.0 changes.
* Always use getcurrenttime() instead of time().
* Support for defining holidays as non-working days in alerts and
  SLA calculations.
* Hosts which appear on multiple pages in the web display can
  use any page they are on in the alerting rules and elsewhere.
* Worker modules (RRD, client-data parsers etc) can operate on
  remote hosts from the hobbitd daemon, for load-sharing.
* Various bugfixes collected over time.

Network test changes:
* Merged new network tests from trunk: SOAP-over-HTTP,
  SSL minimum cipher strength
* Changed network test code to always report a validity period
  for network tests, so it it possible to run network tests less
  often than every 30 minutes (e.g. once an hour).
* Make the content-type setting in HTTP POST tests configurable.
* Make the source-address used for TCP tests configurable.
* Make the acceptable HTTP result codes configurable.
* Use and save HTTP session cookies.

Web changes
* Support generic drop-down lists in templates.
* "NOCOLUMNS" changed to work for all columns.
* New "group-sorted" definition to auto-sort hosts in a group
* Use browser tooltips for host comments
* "Compact" status allows several statuses to appear as a single
  status on the overview webpages.
* Trends page can select the time period to show. Buttons provided
  for the common selections.
* Ghost list report now lists possible candidates for a ghost, 
  based on IP-address or unqualified hostname.

Report changes
* Number of outages as SLA parameter

Miscellaneous
* hobbitlaunch support for running tasks only on certain hosts,
  and for a maximum time.
* Alert script get a unique ID for each alert.



Changes from 4.2.2 -> 4.2.3 (09 Feb 2008)
=========================================
* Time-out code changed to use clock_gettime() with CLOCK_MONOTONIC
* Bugfix for hobbitd/hobbitd_worker communication going out-of-sync
  resulting in "garbled data" being logged and worker modules 
  stopping.
* NCV module now works with negative numbers.
* Several bugfixes in DNS lookup code - could lead to crashes when
  performing DNS tests.
* Switch to C-ARES 1.6.0 - drop support for older versions.
* Run more TCP tests in parallel by not waiting for very slow
  connections to complete before starting new ones.
* Added "hostlist" web utility for spreadsheet-reporting of the
  hosts in Hobbit.


Changes from 4.2.0 -> 4.2.2 (01 Dec 2008)
=========================================

Changelog extracted from Subversion.


------------------------------------------------------------------------
r5935 | storner | 2008-11-26 12:57:11 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   A /branches/4.2.0 (from /trunk:5040)

Create branch for maintenance of the 4.2.x release.


------------------------------------------------------------------------
r5936 | storner | 2008-11-26 13:13:50 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/common/bb-hosts.5

The ability to set a DOWNTIME for an individual test was mentioned in the Changes file, but not documented in the bb-hosts man-page.


------------------------------------------------------------------------
r5937 | storner | 2008-11-26 13:14:48 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitd_client.c

The hobbitd_client program would crash when running in test-mode, i.e. launched with the --test option to test disk-, procs- and ports-settings. This only applies to running it in test-mode, normal operation is not affected.


------------------------------------------------------------------------
r5938 | storner | 2008-11-26 13:15:40 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/lib/headfoot.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/bb-eventlog.c
   M /branches/4.2.0/web/bb-findhost.c
   M /branches/4.2.0/web/bb-rep.c
   M /branches/4.2.0/web/bb-snapshot.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-nkedit.c

Some of the web page would generate an extra line with "Content-type: text/html" at the top of the webpage, although this would most often be hidden by the menubar.


------------------------------------------------------------------------
r5939 | storner | 2008-11-26 13:16:21 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/runclient.sh

Using extra options with the client-side "runclient.sh" script when performing a restart command fails to pass the extra options to the final command. 


------------------------------------------------------------------------

xymon-test/web/bb-webpage.c
/*----------------------------------------------------------------------------*/
/* Hobbit webpage generator tool.                                             */
/*                                                                            */
/* This is a generic webpage generator, that allows scripts to output a       */
/* standard Hobbit-like webpage without having to deal with headers and       */
/* footers.                                                                   */
/*                                                                            */
/* Copyright (C) 2004-2009 Henrik Storner <henrik@storner.dk>                 */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id: bb-webpage.c 6125 2009-02-12 13:09:34Z storner $";

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "libbbgen.h"
#include "version.h"

char *reqenv[] = {
	"BBHOME",
	NULL
};

int main(int argc, char *argv[])
{
	int argi;
	char *hffile = "bb";
	int bgcolor = COL_BLUE;
	char inbuf[8192];
	int n;
	char *envarea = NULL;

	for (argi = 1; (argi < argc); argi++) {
		if (argnmatch(argv[argi], "--env=")) {
			char *p = strchr(argv[argi], '=');
			loadenv(p+1, envarea);
		}
		else if (argnmatch(argv[argi], "--area=")) {
			char *p = strchr(argv[argi], '=');
			envarea = strdup(p+1);
		}
		else if (strcmp(argv[argi], "--debug") == 0) {
			debug = 1;
		}
		else if (argnmatch(argv[argi], "--hffile=")) {
			char *p = strchr(argv[argi], '=');
			hffile = strdup(p+1);
		}
		else if (argnmatch(argv[argi], "--color=")) {
			char *p = strchr(argv[argi], '=');
			bgcolor = parse_color(p+1);
		}
	}

	envcheck(reqenv);

	fprintf(stdout, "Content-type: %s\n\n", xgetenv("HTMLCONTENTTYPE"));
	
	headfoot(stdout, hffile, "", "header", bgcolor);
	do {
		n = fread(inbuf, 1, sizeof(inbuf), stdin);
		if (n > 0) fwrite(inbuf, 1, n, stdout);
	} while (n == sizeof(inbuf));
	headfoot(stdout, hffile, "", "footer", bgcolor);

	return 0;
}

